File: challenges/programming/esolang/interpreters/py/ctf/world.py
import string
from .player import Player
from .enums import Tile

from typing import List, Dict, Tuple, Set

class World:
    __grid: List[List[int]]
    __drops: Set[Tuple[int, int]]
    __update_queue: Dict[Tuple[int, int], int]
    __output: List[int]
    __input: List[int]
    new_players: List['Player']

    def __init__(self, grid: List[List[int]]):
        self.__grid = grid

        width = len(max(self.__grid, key=lambda x: len(x)))
        for row in self.__grid:
            row += [Tile.EMPTY.value] * width

        self.__drops = set()
        self.new_players = []
        self.__update_queue = {}
        self.__output = []
        self.__input = []

        for y, row in enumerate(grid):
            for x, c in enumerate(row):
                if c == Tile.DROP.value:
                    self.__drops.add((x, y))

    @property
    def input(self) -> bytes:
        return bytes(self.__input)

    @property
    def output(self) -> bytes:
        return bytes(self.__output)

    def get_tile(self, position: Tuple[int, int]) -> int:
        tile = self.__grid[position[1]][position[0]]

        return tile

    def put_tile(self, position: Tuple[int, int], tile: int):
        if tile == Tile.DROP.value:
            self.__drops.add(position)

        self.__update_queue[position] = tile
    
    def pop_tile(self, position: Tuple[int, int]):
        tile = self.get_tile(position)

        if self.__is_drop(position):
            self.delete_tile(position)

        return tile

    def delete_tile(self, position: Tuple[int, int]):
        self.put_tile(position, Tile.EMPTY.value)
    
    def step(self):
        self.__update_tiles()

    def raw(self) -> List[List[int]]:
        return [[c for c in row] for row in self.__grid]

    def print(self, c: int):
        self.__output.append(c)

    def push_input(self):
        inp = input("Enter: ").strip()

        if len(inp) == 0:
            self.__input.append(ord('0'))
            return

        self.__input += list(c for c in reversed(inp.encode('utf8')))

    def pop_input(self) -> int:
        return self.__input.pop()

    def input_empty(self) -> bool:
        return len(self.__input) == 0

    def __update_tiles(self):
        for position, tile in self.__update_queue.items():
            if tile == Tile.EMPTY.value and self.__is_drop(position):
                tile = Tile.DROP.value

            self.__set_tile(position, tile)

        self.__update_queue = {}
    
    def __is_drop(self, position: Tuple[int, int]) -> bool:
        return position in self.__drops

    def __set_tile(self, position: Tuple[int, int], val: int):
        self.__grid[position[1]][position[0]] = val

File: challenges/programming/esolang/interpreters/js/src/world.ts
import { Position, TileType, Tile, PositionIndex } from './utils';

function longestRow(grid: any[][]): number {
    let maxLength = 0;

    for(let row of grid) {
        maxLength = Math.max(maxLength, row.length);
    }

    return maxLength;
}

export class World {
    private __grid: TileType[][];
    private __drops: Set<PositionIndex>;
    private __updateQueue: { [key: string]: TileType };
    
    constructor(grid: TileType[][]) {
        this.__updateQueue = {};
        this.__grid = grid;

        this.__drops = new Set<PositionIndex>();
        this.__grid.forEach((row, y) => row.forEach((tile, x) => {
            if(tile == Tile.DROP) {
                this.__drops.add(new Position(x, y).toString());
            }
        }))

        let gridWidth = longestRow(this.__grid);
        this.__grid = this.__grid.map(row => [...row, ...new Array(gridWidth - row.length).fill(Tile.EMPTY)]);
    }

    public raw() {
        return this.__grid.map(row => row.map(tile => tile));
    }

    public step() {
        this.updateTiles();
    }

    public getTile(position: Position): TileType {
        return this.__grid[position.y][position.x];
    }

    public putTile(position: Position, tile: TileType) {
        if(tile == Tile.DROP) {
            this.__drops.add(position.toString());
        }

        this.__updateQueue[position.toString()] = tile;
    }

    public popTile(position: Position): TileType {
        let tile = this.getTile(position);

        if(this.isDrop(position)) this.deleteTile(position);

        return tile;
    }

    public deleteTile(position: Position) {
        this.putTile(position, Tile.EMPTY);
    }

    private updateTiles() {
        Object.entries(this.__updateQueue).forEach(([positionIndex, tile]) => {
            let position = Position.from(positionIndex);

            if(tile == Tile.EMPTY && this.isDrop(position)) tile = Tile.DROP;

            this.setTile(position, tile as TileType);
        });

        this.__updateQueue = {};
    }

    private setTile(position: Position, tile: TileType) {
        this.__grid[position.y][position.x] = tile;
    }

    private isDrop(position: Position): boolean {
        return this.__drops.has(position.toString());
    }
}
